lib/client/db/dbcmd/dbcmd.go

```diff
@@ -464,16 +464,25 @@ func (c *CLICommandBuilder) isSqlcmdAvailable() bool {
 	return c.isBinAvailable(sqlcmdBin)
 }
 
-func (c *CLICommandBuilder) shouldUseMongoshBin() bool {
+func (c *CLICommandBuilder) shouldUseMongoshBin(db types.Database) bool {
+	// DocumentDB prefers the legacy "mongo" client.
+	if db.GetType() == types.DatabaseTypeDocumentDB {
+		return c.isMongoshBinAvailable() && !c.isBinAvailable(mongoBin)
+	}
 	// Use "mongosh" if available.
 	// If not, use legacy "mongo" if available.
 	// If both are not available, pick "mongosh" in print out.
 	return c.isMongoshBinAvailable() || !c.isBinAvailable(mongoBin)
 }
 
-func (c *CLICommandBuilder) getMongoCommand() *exec.Cmd {
+func (c *CLICommandBuilder) getMongoCommand(ctx context.Context) (*exec.Cmd, error) {
+	db, err := c.getDatabase(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
This approach will make a backend call where the database object should already be available.

Instead of injecting a function to retrieve the database object, we can inject information about the database object directly and save the backend call:

In Teleport, where the database gateway is created:
https://github.com/gravitational/teleport/blob/b32bd452bc18c68f1c716ecbaa99a14a79e2cb8b/lib/teleterm/clusters/cluster_gateways.go#L75

In the tsh flow:
https://github.com/gravitational/teleport/blob/b415c2d44a8119b80d9d1895130c69ac48106759/tool/tsh/common/proxy.go#L159

Does it make sense to just inject the option to the builder from the caller like `WithDefaultMongoClient(...)` instead of making this call deep inside the CLI builder?

Reply:
> This approach will make a backend call where the database object should already be available.

For teleterm gateway yes it's an extra call. For `tsh db connect`, it's already cached. 

> Does it make sense to just inject the option to the builder from the caller like WithDefaultMongoClient(...) instead of making this call deep inside the CLI builder?

I can do this, but I will have to attach a `types.Database` to the teleterm gateway (which affects several other layers on teleterm).

Let me see if i can find a better solution.

Reply:
I took another look but i don't think `WithDefaultMongoClient` will work as there is another check in dbcmd for DocumentDB to add the `--retryWrites=false` check. I can probably do `WithDatabaseType`?

Also, looking back why I did it this way, I actually prefer to keep these logic inside `dbcmd` instead of doing several if checks on the outside:
https://github.com/gravitational/teleport/blob/f690c75f438c6412a4842922e71b8f41375935a7/tool/tsh/common/db.go#L779-L784
As we have to copy the outside logic in many places (and we are still missing some of these checks for teleterm code).back

Let me know what you think about this. I propose keeping the `GetDatabase` callback but I will make sure it uses a cached `types.Database` for teleterm. But I am open to other suggestions.

Reply:
All right, If this require a lot of work I don't mind sipping current version.